File: atcoder/abc294/g.rs
#![allow(unused)]

fn main() {
    let n = read::<usize>();
    let mut adj = vec![vec![]; n];
    let mut edges = vec![];
    for _ in 0..n - 1 {
        let inp = readv::<usize>();
        let (u, v, w) = (inp[0] - 1, inp[1] - 1, inp[2] as i64);
        adj[u].push(v);
        adj[v].push(u);
        edges.push((u, v, w));
    }

    let hld = HLD::new(&adj, 0);

    let mut bit = Bit::<i64>::new(n + 1);
    for &(u, v, w) in edges.iter() {
        let (u, v) = if hld.parent[u] == v { (v, u) } else { (u, v) };
        bit.add(hld.t_enter[v], w);
        bit.add(hld.t_leave[v], -w);
    }

    let q = read::<usize>();
    let mut ans = vec![];
    for _ in 0..q {
        let inp = readv::<usize>();
        let (c, a, b) = (inp[0], inp[1], inp[2]);

        if c == 1 {
            let (u, v, w) = edges[a - 1];
            let (u, v) = if hld.parent[u] == v { (v, u) } else { (u, v) };
            let b = b as i64;
            bit.add(hld.t_enter[v], b - w);
            bit.add(hld.t_leave[v], -b + w);
            edges[a - 1] = (u, v, b);
        } else {
            let (a, b) = (a - 1, b - 1);
            let lca = hld.lca(a, b);
            let da = bit.sum(0, hld.t_enter[a] + 1);
            let db = bit.sum(0, hld.t_enter[b] + 1);
            let dl = bit.sum(0, hld.t_enter[lca] + 1);
            let v = da + db - 2 * dl;
            ans.push(v);
        }
    }

    println!("{}", join(&ans, "\n"));
}

#[derive(Debug)]
struct HLD {
    size: Vec<usize>,
    depth: Vec<usize>,
    heavy: Vec<usize>,
    parent: Vec<usize>,

    top: Vec<usize>,
    tour: Vec<usize>,
    t_enter: Vec<usize>,
    t_leave: Vec<usize>,
}

impl HLD {
    fn new(adj: &Vec<Vec<usize>>, root: usize) -> Self {
        let n = adj.len();

        // DFS to find size, depth, heavy, parent
        let mut size = vec![1; n];
        let mut depth = vec![0; n];
        let mut heavy = vec![std::usize::MAX; n];
        let mut parent = vec![std::usize::MAX; n];
        let mut stack = vec![('l', root), ('e', root)];
        depth[root] = 0;
        parent[root] = root;
        while let Some((cmd, u)) = stack.pop() {
            if cmd == 'l' {
                size[u] = 1;
                heavy[u] = std::usize::MAX;
                for &v in adj[u].iter().rev() {
                    if v != parent[u] {
                        size[u] += size[v];
                        if heavy[u] == std::usize::MAX || size[v] > size[heavy[u]] {
                            heavy[u] = v;
                        }
                    }
                }
            } else {
                for &v in adj[u].iter().rev() {
                    if parent[v] == std::usize::MAX {
                        parent[v] = u;
                        depth[v] = depth[u] + 1;
                        stack.push(('l', v));
                        stack.push(('e', v));
                    }
                }
            }
        }

        // DFS to decompose tree
        let mut time = 0;
        let mut top = vec![std::usize::MAX; n];
        let mut tour = vec![std::usize::MAX; n];
        let mut t_enter = vec![std::usize::MAX; n];
        let mut t_leave = vec![std::usize::MAX; n];
        let mut stack = vec![('l', root), ('e', root)];
        top[root] = root;
        while let Some((cmd, u)) = stack.pop() {
            if cmd == 'l' {
                t_leave[u] = time;
            } else {
                t_enter[u] = time;
                tour[time] = u;
                time += 1;
                for &v in adj[u].iter().rev() {
                    if v != parent[u] {
                        top[v] = if v == heavy[u] { top[u] } else { v };
                        stack.push(('l', v));
                        stack.push(('e', v));
                    }
                }
            }
        }

        Self {
            size,
            depth,
            heavy,
            parent,
            top,
            tour,
            t_enter,
            t_leave,
        }
    }

    fn lca(&self, mut u: usize, mut v: usize) -> usize {
        assert!(u < self.top.len());
        assert!(v < self.top.len());
        // Until they are in the same path
        while self.top[u] != self.top[v] {
            if self.depth[self.top[u]] > self.depth[self.top[v]] {
                u = self.parent[self.top[u]];
            } else {
                v = self.parent[self.top[v]];
            }
        }
        if self.depth[u] < self.depth[v] {
            u
        } else {
            v
        }
    }
}

struct Bit<T> {
    data: Vec<T>,
}

impl<T> Bit<T>
where
    T: Copy + Default + std::ops::Add<Output = T> + std::ops::Sub<Output = T>,
{
    fn new(n: usize) -> Self {
        Self {
            data: vec![T::default(); n + 1],
        }
    }

    fn add(&mut self, idx: usize, val: T) {
        let n = self.data.len() - 1;
        assert!(idx < n);
        let mut i = idx + 1;
        while i <= n {
            self.data[i] = self.data[i] + val;
            i += i & (!i + 1);
        }
    }

    fn prefix(&self, idx: usize) -> T {
        let n = self.data.len() - 1;
        assert!(idx < n);
        let mut res = T::default();
        let mut i = idx + 1;
        while i > 0 {
            res = res + self.data[i];
            i -= i & (!i + 1);
        }
        res
    }

    // l..r
    fn sum(&self, l: usize, r: usize) -> T {
        let n = self.data.len() - 1;
        assert!(l < r && r <= n);
        let val = self.prefix(r - 1);
        if l == 0 {
            val
        } else {
            val - self.prefix(l - 1)
        }
    }
}

fn read<T: std::str::FromStr>() -> T {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).ok();
    s.trim().parse().ok().unwrap()
}

fn readv<T: std::str::FromStr>() -> Vec<T> {
    read::<String>()
        .split_ascii_whitespace()
        .map(|t| t.parse().ok().unwrap())
        .collect()
}

fn reads() -> Vec<char> {
    read::<String>().chars().collect::<Vec<char>>()
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}

File: atcoder/abc294/g.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>
using namespace std;

struct HLD {
    vector<int> size;
    vector<int> depth;
    vector<int> heavy;
    vector<int> parent;
    vector<int> top;
    vector<int> tour;
    vector<int> t_enter;
    vector<int> t_leave;

    HLD(const vector<vector<int>> &adj, int root) {
        int N = adj.size();
        const int inf = 0x3f3f3f3f;

        // Find size, depth, heavy, parent by DFS
        {
            size = vector<int>(N, 1);
            depth = vector<int>(N, 0);
            heavy = vector<int>(N, inf);
            parent = vector<int>(N, inf);
            auto stack = vector<pair<char, int>>{{'l', root}, {'e', root}};
            depth[root] = 0;
            parent[root] = root;
            while (!stack.empty()) {
                auto [cmd, u] = stack.back();
                stack.pop_back();
                if (cmd == 'l') {
                    size[u] = 1;
                    heavy[u] = inf;
                    for (auto v : adj[u]) {
                        if (v != parent[u]) {
                            size[u] += size[v];
                            if (heavy[u] == inf || size[v] > size[heavy[u]]) {
                                heavy[u] = v;
                            }
                        }
                    }
                } else {
                    for (auto v : adj[u]) {
                        if (parent[v] == inf) {
                            parent[v] = u;
                            depth[v] = depth[u] + 1;
                            stack.push_back({'l', v});
                            stack.push_back({'e', v});
                        }
                    }
                }
            }
        }

        // Find top, tour, t_enter, t_leave by DFS
        {
            int time = 0;
            top = vector<int>(N, inf);
            tour = vector<int>(N, inf);
            t_enter = vector<int>(N, inf);
            t_leave = vector<int>(N, inf);
            auto stack = vector<pair<char, int>>{{'l', root}, {'e', root}};
            top[root] = root;
            while (!stack.empty()) {
                auto [cmd, u] = stack.back();
                stack.pop_back();
                if (cmd == 'l') {
                    t_leave[u] = time;
                } else {
                    t_enter[u] = time;
                    tour[time] = u;
                    time++;
                    for (auto v : adj[u]) {
                        if (v != parent[u]) {
                            // heavy edge / light edge
                            top[v] = ((v == heavy[u]) ? top[u] : v);
                            stack.push_back({'l', v});
                            stack.push_back({'e', v});
                        }
                    }
                }
            }
        }
    }

    int lca(int u, int v) {
        while (top[u] != top[v]) {
            if (depth[top[u]] > depth[top[v]]) {
                u = parent[top[u]];
            } else {
                v = parent[top[v]];
            }
        }
        return ((depth[u] < depth[v]) ? u : v);
    }
};

template <typename T> struct BIT {
    vector<T> data;

    BIT(int n) { data = vector<T>(n, 0); }

    void add(int idx, T val) {
        for (int i = idx + 1; i <= data.size() - 1; i += (i & (-i))) {
            data[i] = data[i] + val;
        }
    }

    T prefix(int idx) {
        T res = 0;
        for (int i = idx + 1; i > 0; i -= (i & -i)) {
            res = res + data[i];
        }
        return res;
    }

    T sum(int l, int r) { // l..r
        T val = prefix(r - 1);
        if (l != 0) {
            val -= prefix(l - 1);
        }
        return val;
    }
};

using ll = long long;

template <typename T> ostream &operator<<(ostream &out, const vector<T> v) {
    out << "[ ";
    for (auto x : v) {
        out << x << ", ";
    }
    out << "]";
    return out;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    vector<vector<int>> adj(N);
    vector<tuple<int, int, ll>> edges;
    for (int i = 0; i < N - 1; i++) {
        int u, v;
        ll w;
        cin >> u >> v >> w;
        u--;
        v--;
        adj[u].push_back(v);
        adj[v].push_back(u);
        edges.push_back({u, v, w});
    }

    auto hld = HLD(adj, 0);

    auto bit = BIT<ll>(N + 1);
    for (auto [u, v, w] : edges) {
        if (hld.parent[u] == v) {
            swap(u, v);
        }
        bit.add(hld.t_enter[v], +w);
        bit.add(hld.t_leave[v], -w);
    }

    int Q;
    cin >> Q;
    while (Q--) {
        int cmd, a, b;
        cin >> cmd >> a >> b;

        if (cmd == 1) {
            auto [u, v, w] = edges[a - 1];
            if (hld.parent[u] == v) {
                swap(u, v);
            }
            ll nw = b;
            bit.add(hld.t_enter[v], nw - w);
            bit.add(hld.t_leave[v], -nw + w);
            edges[a - 1] = {u, v, nw};
        } else {
            a--;
            b--;
            int lca = hld.lca(a, b);
            ll da = bit.sum(0, hld.t_enter[a] + 1);
            ll db = bit.sum(0, hld.t_enter[b] + 1);
            ll dl = bit.sum(0, hld.t_enter[lca] + 1);
            ll ans = da + db - 2 * dl;
            cout << ans << "\n";
        }
    }

    return 0;
}